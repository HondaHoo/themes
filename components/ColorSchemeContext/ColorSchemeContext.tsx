'use client';

import {
  createContext,
  useState,
  ReactNode,
  Dispatch,
  SetStateAction,
  useReducer,
  Reducer,
  useCallback,
} from 'react';

import {colourShemeAndMeta, Lightness} from '@/types';
import {
  ColorSchemeAction,
  colorSchemeReducer,
  colorSchemeReducerInitialiser,
  ColorSchemeState,
  useDispatchActions,
} from '@/components/ColorSchemeContext/colorSchemeContextReducer';

export const CurrentColorSchemeContext =
  createContext<colourShemeAndMeta['name']>('');
export const SetCurrentColorSchemeContext = createContext<
  (colorSchemeName: colourShemeAndMeta['name']) => void
>(() => {});
export const CurrentLightnessContext = createContext<Lightness>('light');
export const SetCurrentLightnessContext = createContext<
  (lightness: Lightness) => void
>(() => {});
export const SetNextPrevColorSchemeContext = createContext<
  (direction: 'next' | 'prev') => void
>(() => {});

type ColorSchemesProviderProps = {
  children: ReactNode;
  colorSchemes: colourShemeAndMeta[];
};

const initState: ColorSchemeState = {
  currentColorScheme: '',
  currentLightness: 'dark',
  lightColorSchemes: [],
  darkColorSchemes: [],
};

export const ColorSchemesProvider = (props: ColorSchemesProviderProps) => {
  const [state, dispatch] = useReducer(colorSchemeReducer, initState, () =>
    colorSchemeReducerInitialiser(props.colorSchemes)
  );

  const {setCurrentColorScheme, setCurrentLightness, setNextPrevColorScheme} =
    useDispatchActions(dispatch);

  return (
    <CurrentColorSchemeContext.Provider value={state.currentColorScheme}>
      <SetCurrentColorSchemeContext.Provider value={setCurrentColorScheme}>
        <CurrentLightnessContext.Provider value={state.currentLightness}>
          <SetCurrentLightnessContext.Provider value={setCurrentLightness}>
            <SetNextPrevColorSchemeContext.Provider
              value={setNextPrevColorScheme}
            >
              {props.children}
            </SetNextPrevColorSchemeContext.Provider>
          </SetCurrentLightnessContext.Provider>
        </CurrentLightnessContext.Provider>
      </SetCurrentColorSchemeContext.Provider>
    </CurrentColorSchemeContext.Provider>
  );
};
