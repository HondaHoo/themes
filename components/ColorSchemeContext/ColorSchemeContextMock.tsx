// this is a helper file to allow Playwright to test the ColorSchemeContext
// see: https://playwright.dev/docs/next/test-components#under-the-hood
'use client';

import {useContext} from 'react';

import {
  CurrentColorSchemeContext,
  SetCurrentColorSchemeContext,
} from './ColorSchemeContext';

type Props = {
  nextColorScheme: string;
};

const ColorSchemeContextMock = (props: Props) => {
  const colorScheme = useContext(CurrentColorSchemeContext);
  const setCurrentColorSchemeName = useContext(SetCurrentColorSchemeContext);
  return (
    <>
      <h1>{colorScheme.name}</h1>
      <button
        onClick={() => {
          setCurrentColorSchemeName(props.nextColorScheme);
        }}
      >
        Change name
      </button>
    </>
  );
};

export default ColorSchemeContextMock;
