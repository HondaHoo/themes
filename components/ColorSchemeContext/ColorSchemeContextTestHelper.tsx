// this is a helper file to allow Playwright to test the ColorSchemeContext
// see: https://playwright.dev/docs/next/test-components#under-the-hood
'use client';

import {useContext} from 'react';

import {
  CurrentColorSchemeContext,
  SetCurrentColorSchemeContext,
} from './ColorSchemeContext';

const ConsumerTestComponent = () => {
  const colorSchemeName = useContext(CurrentColorSchemeContext);
  const setCurrentColorSchemeName = useContext(SetCurrentColorSchemeContext);
  return (
    <>
      <h1>{colorSchemeName}</h1>
      <button
        onClick={() => {
          setCurrentColorSchemeName('Solarized Dark');
        }}
      >
        Change name
      </button>
    </>
  );
};

export default ConsumerTestComponent;
