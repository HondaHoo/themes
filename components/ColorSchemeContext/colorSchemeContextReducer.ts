import {colorSchemeAndMeta, Lightness} from '@/types';
import immer from 'immer';
import {Dispatch, useCallback} from 'react';

export type ColorSchemeState = {
  currentColorScheme: colorSchemeAndMeta;
  currentLightness: Lightness;
  lightColorSchemes: colorSchemeAndMeta[];
  darkColorSchemes: colorSchemeAndMeta[];
  colorSchemes: colorSchemeAndMeta[];
};

export type ColorSchemeAction =
  | {
      type: 'setColorScheme';
      payload: {
        colorSchemeName: string;
      };
    }
  | {
      type: 'setNextPrevColorScheme';
      payload: {
        direction: 'next' | 'prev';
      };
    }
  | {
      type: 'setLightness';
      payload: {
        lightness: Lightness;
      };
    };

export const colorSchemeReducer = (
  state: ColorSchemeState,
  action: ColorSchemeAction
): ColorSchemeState =>
  immer(state, (draft: ColorSchemeState) => {
    switch (action.type) {
      case 'setColorScheme':
        {
          const currentColorScheme = state.colorSchemes.find(
            (colorScheme) => colorScheme.name === action.payload.colorSchemeName
          );
          if (!currentColorScheme) {
            throw new Error(
              `Color scheme ${action.payload.colorSchemeName} not found`
            );
          }
          draft.currentColorScheme = currentColorScheme;
        }
        break;
      case 'setLightness':
        {
          draft.currentLightness = action.payload.lightness;
          // reset current color scheme to first in lightness
          const currentColorSchemesByLightness =
            draft.currentLightness === 'light'
              ? draft.lightColorSchemes
              : draft.darkColorSchemes;
          draft.currentColorScheme = currentColorSchemesByLightness[0];
        }
        break;
      case 'setNextPrevColorScheme':
        {
          const currentColorSchemesByLightness =
            draft.currentLightness === 'light'
              ? draft.lightColorSchemes
              : draft.darkColorSchemes;
          draft.currentColorScheme = getNextPrevColorScheme(
            currentColorSchemesByLightness,
            draft.currentColorScheme.name,
            action.payload.direction
          );
        }
        break;
    }
  });

export const colorSchemeReducerInitialiser = (
  colorSchemes: colorSchemeAndMeta[]
): ColorSchemeState => {
  const lightColorSchemes = colorSchemesFilteredByCurrentLightness(
    colorSchemes,
    false
  );
  const darkColorSchemes = colorSchemesFilteredByCurrentLightness(
    colorSchemes,
    true
  );

  return {
    currentColorScheme: darkColorSchemes[0],
    currentLightness: 'dark',
    lightColorSchemes,
    darkColorSchemes,
    colorSchemes,
  };
};

export const useDispatchActions = (dispatch: Dispatch<ColorSchemeAction>) => {
  const setCurrentColorScheme = useCallback(
    (colorSchemeName: colorSchemeAndMeta['name']) => {
      dispatch({
        type: 'setColorScheme',
        payload: {
          colorSchemeName,
        },
      });
    },
    []
  );

  const setCurrentLightness = useCallback((lightness: Lightness) => {
    dispatch({
      type: 'setLightness',
      payload: {
        lightness,
      },
    });
  }, []);

  const setNextPrevColorScheme = useCallback((direction: 'next' | 'prev') => {
    dispatch({
      type: 'setNextPrevColorScheme',
      payload: {
        direction,
      },
    });
  }, []);

  return {
    setCurrentColorScheme,
    setCurrentLightness,
    setNextPrevColorScheme,
  };
};

export const getNextPrevColorScheme = (
  colorSchemes: colorSchemeAndMeta[],
  currentColorScheme: string,
  direction: 'next' | 'prev'
) => {
  const currentIndex = colorSchemes.findIndex(
    (colorScheme) => colorScheme.name === currentColorScheme
  );
  if (currentIndex === -1) {
    throw new Error('currentColorScheme not found in colorSchemes');
  }
  const nextIndex =
    direction === 'next'
      ? (currentIndex + 1) % colorSchemes.length
      : (currentIndex - 1 + colorSchemes.length) % colorSchemes.length;
  return colorSchemes[nextIndex];
};

export const colorSchemesFilteredByCurrentLightness = (
  colorSchemes: colorSchemeAndMeta[],
  lightnessIsDark: boolean
) => {
  return colorSchemes.filter(
    (colorScheme) => colorScheme.meta.isDark === lightnessIsDark
  );
};
