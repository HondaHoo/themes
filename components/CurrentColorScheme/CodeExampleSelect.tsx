'use client';

import {
  Root as TabRoot,
  List as TabList,
  Trigger as TabTrigger,
  Content as TabContent,
} from '@radix-ui/react-tabs';

import ChalkExample from '@/components/CurrentColorScheme/CodeExamples/ChalkExample';
import JestSummary from '@/components/CurrentColorScheme/CodeExamples/JestSummary';

const examplesEnum = [
  {
    id: 'jest',
    label: 'Jest',
    component: JestSummary,
  },
  {
    id: 'chalk',
    label: 'Chalk',
    component: ChalkExample,
  },
] as const;

const CodeExampleSelect = () => {
  return (
    <TabRoot defaultValue={examplesEnum[0].id}>
      <TabList aria-label="Code example tabs">
        {examplesEnum.map(({id, label}) => (
          <TabTrigger key={id} value={id}>
            {label}
          </TabTrigger>
        ))}
      </TabList>
      {examplesEnum.map(({id, component: Component}) => (
        <TabContent key={id} value={id}>
          <Component />
        </TabContent>
      ))}
    </TabRoot>
  );
};

export default CodeExampleSelect;
