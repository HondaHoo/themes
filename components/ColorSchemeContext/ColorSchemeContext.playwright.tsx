import {test, expect} from '@playwright/experimental-ct-react';
import ColorSchemeContextMock from './ColorSchemeContextMock';

const themes = [
  {
    name: 'Dark theme',
    meta: {
      isDark: true,
    },
  },
  {
    name: 'Solarized Dark',
    meta: {
      isDark: true,
    },
  },
];

test('should show first theme by default', async ({mount}) => {
  const component = await mount(<ColorSchemeContextMock nextColorScheme="" />, {
    hooksConfig: {colorSchemes: themes},
  });
  await expect(component.getByRole('heading', {level: 1})).toHaveText(
    'Dark theme'
  );
});

test('should change theme', async ({mount}) => {
  const component = await mount(
    <ColorSchemeContextMock nextColorScheme="Solarized Dark" />,
    {
      hooksConfig: {colorSchemes: themes},
    }
  );
  await expect(component.getByRole('heading', {level: 1})).toHaveText(
    'Dark theme'
  );
  await component.getByRole('button').click();
  await expect(component.getByRole('heading', {level: 1})).toHaveText(
    'Solarized Dark'
  );
});
