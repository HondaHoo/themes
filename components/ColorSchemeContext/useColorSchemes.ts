import {
  ColorSchemeStateContext,
  SetColorSchemeStateContext,
} from '@/components/ColorSchemeContext/ColorSchemeContext';
import {useDispatchActions} from '@/components/ColorSchemeContext/colorSchemeContextReducer';

import useDefinedContext from '@/utilities/useDefinedContext';

const useColorSchemes = () => {
  const colorSchemeState = useDefinedContext(ColorSchemeStateContext);
  const setColorSchemeDispatch = useDefinedContext(SetColorSchemeStateContext);

  const {setActiveColorScheme, setLightness, setNextPrevColorScheme} =
    useDispatchActions(setColorSchemeDispatch);

  return {
    colorSchemeState,
    setActiveColorScheme,
    setLightness,
    setNextPrevColorScheme,
  };
};

export default useColorSchemes;
