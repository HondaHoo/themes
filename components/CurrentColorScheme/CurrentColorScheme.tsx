// display all the properties of the current color scheme, with a box next to the colours with that colour
// Compare this snippet from components/ColorSchemeSelector/ColorSchemeSelector.tsx:
// // react component, displays all the color schemes
'use client';
import {CSSProperties} from 'react';

import {CurrentColorSchemeContext} from '@/components/ColorSchemeContext/ColorSchemeContext';
import {colorSchemeAndMeta} from '@/types';
import useDefinedContext from '@/utilities/useDefinedContext';
import css from './CurrentColorScheme.module.css';
import ChalkExample from '@/components/CurrentColorScheme/CodeExamples/ChalkExample';

const getColorsOnlyFromSchemeAndMeta = (colorScheme: colorSchemeAndMeta) => {
  const {name, meta, ...colorSchemeColors} = colorScheme;
  return colorSchemeColors;
};

interface StyleProps extends CSSProperties {
  '--black': string;
  '--red': string;
  '--green': string;
  '--yellow': string;
  '--blue': string;
  '--purple': string;
  '--cyan': string;
  '--white': string;
  '--brightBlack': string;
  '--brightRed': string;
  '--brightGreen': string;
  '--brightYellow': string;
  '--brightBlue': string;
  '--brightPurple': string;
  '--brightCyan': string;
  '--brightWhite': string;
  '--background': string;
  '--foreground': string;
  '--cursorColor'?: string;
  '--selectionBackground'?: string;
}

const getColorsForCssVars = (colorScheme: colorSchemeAndMeta): StyleProps => {
  return {
    '--black': colorScheme.black,
    '--red': colorScheme.red,
    '--green': colorScheme.green,
    '--yellow': colorScheme.yellow,
    '--blue': colorScheme.blue,
    '--purple': colorScheme.purple,
    '--cyan': colorScheme.cyan,
    '--white': colorScheme.white,
    '--brightBlack': colorScheme.brightBlack,
    '--brightRed': colorScheme.brightRed,
    '--brightGreen': colorScheme.brightGreen,
    '--brightYellow': colorScheme.brightYellow,
    '--brightBlue': colorScheme.brightBlue,
    '--brightPurple': colorScheme.brightPurple,
    '--brightCyan': colorScheme.brightCyan,
    '--brightWhite': colorScheme.brightWhite,
    '--background': colorScheme.background,
    '--foreground': colorScheme.foreground,
    '--cursorColor': colorScheme.cursorColor,
    '--selectionBackground': colorScheme.selectionBackground,
  };
};

const CurrentColorScheme = () => {
  const currentColorScheme = useDefinedContext(CurrentColorSchemeContext);
  const currentColors = getColorsOnlyFromSchemeAndMeta(currentColorScheme);
  const colorsForCssVars = getColorsForCssVars(currentColorScheme);

  return (
    <div className={css.container} style={colorsForCssVars}>
      <h2>{currentColorScheme.name}</h2>
      <ChalkExample />
      <ul>
        {Object.entries(currentColors).map(([key, value]) => (
          <li key={key}>
            {key}: {value}
            <div
              className={css[key]}
              style={{
                width: '100px',
                height: '1rem',
                display: 'inline-block',
              }}
            ></div>
          </li>
        ))}
      </ul>
    </div>
  );
};

export default CurrentColorScheme;
