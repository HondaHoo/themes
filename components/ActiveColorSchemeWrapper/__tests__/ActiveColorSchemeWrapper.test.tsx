import {render, screen} from '@testing-library/react';
import '@testing-library/jest-dom';

import CurrentColorScheme from '../ActiveColorSchemeWrapper';

import {
  ColorSchemeStateContext,
  SetColorSchemeStateContext,
} from '@/components/ColorSchemeContext/ColorSchemeContext';
import {ColorSchemeState} from '@/components/ColorSchemeContext/colorSchemeContextReducer';
import useColorSchemes from '@/components/ColorSchemeContext/useColorSchemes';
import {lightThemeA} from '@/utilities/mockColorSchemes';
import getColorsForCssVars from '@/utilities/getColorsForCssVars';

const colorSchemeState: ColorSchemeState = {
  activeColorScheme: lightThemeA,
  lightness: 'light',
  lightColorSchemes: [lightThemeA],
  darkColorSchemes: [],
  colorSchemes: [lightThemeA],
};

const dispatch = jest.fn();

const MockActiveColorSchemeName = () => {
  const {
    colorSchemeState: {activeColorScheme},
  } = useColorSchemes();
  const cssVarColors = getColorsForCssVars(activeColorScheme);
  const colors = Object.entries(cssVarColors);
  return (
    <ul style={cssVarColors}>
      {colors.map(([key, value], i) => {
        return (
          <li key={i}>
            {key} == {value}
          </li>
        );
      })}
    </ul>
  );
};

it('should render the current color scheme', () => {
  render(
    <ColorSchemeStateContext.Provider value={colorSchemeState}>
      <SetColorSchemeStateContext.Provider value={dispatch}>
        <CurrentColorScheme>
          <MockActiveColorSchemeName />
        </CurrentColorScheme>
      </SetColorSchemeStateContext.Provider>
    </ColorSchemeStateContext.Provider>
  );

  // Jest uses JS Dom which doesn't support CSS Variables, so this is fairly hollow test
  expect(screen.getByRole('list')).toMatchSnapshot();
});
