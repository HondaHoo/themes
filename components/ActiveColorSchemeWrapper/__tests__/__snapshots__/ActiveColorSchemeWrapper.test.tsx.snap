// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the current color scheme 1`] = `
<ul
  style="--black: #090300; --red: #db2d20; --green: #01a252; --yellow: #fded02; --blue: #01a0e4; --purple: #a16a94; --cyan: #b5e4f4; --white: #a5a2a2; --brightBlack: #5c5855; --brightRed: #e8bbd0; --brightGreen: #3a3432; --brightYellow: #4a4543; --brightBlue: #807d7c; --brightPurple: #d6d5d4; --brightCyan: #cdab53; --brightWhite: #f7f7f7; --background: #f7f7f7; --foreground: #4a4543; --cursorColor: #4a4543; --selectionBackground: #a5a2a2;"
>
  <li>
    --black
     == 
    #090300
  </li>
  <li>
    --red
     == 
    #db2d20
  </li>
  <li>
    --green
     == 
    #01a252
  </li>
  <li>
    --yellow
     == 
    #fded02
  </li>
  <li>
    --blue
     == 
    #01a0e4
  </li>
  <li>
    --purple
     == 
    #a16a94
  </li>
  <li>
    --cyan
     == 
    #b5e4f4
  </li>
  <li>
    --white
     == 
    #a5a2a2
  </li>
  <li>
    --brightBlack
     == 
    #5c5855
  </li>
  <li>
    --brightRed
     == 
    #e8bbd0
  </li>
  <li>
    --brightGreen
     == 
    #3a3432
  </li>
  <li>
    --brightYellow
     == 
    #4a4543
  </li>
  <li>
    --brightBlue
     == 
    #807d7c
  </li>
  <li>
    --brightPurple
     == 
    #d6d5d4
  </li>
  <li>
    --brightCyan
     == 
    #cdab53
  </li>
  <li>
    --brightWhite
     == 
    #f7f7f7
  </li>
  <li>
    --background
     == 
    #f7f7f7
  </li>
  <li>
    --foreground
     == 
    #4a4543
  </li>
  <li>
    --cursorColor
     == 
    #4a4543
  </li>
  <li>
    --selectionBackground
     == 
    #a5a2a2
  </li>
</ul>
`;
